src/lib.rs

```diff
@@ -295,70 +295,81 @@ fn parse_unary_atomic_expression(input: ParseInput) -> ParseResult<Expression> {
 ///
 /// parse_infix_expression handles any and all kinds of expressions that would
 /// left-recursively call parse_expression
-fn parse_infix_expression(input: ParseInput, left: Expression) -> ParseResult<Expression> {
+fn parse_infix_expression<'a>(
+    left: Expression,
+) -> impl FnMut(ParseInput<'a>) -> ParseResult<Expression> {
     // this implementation follows the 'Top Down Operator Precedence' algorithm. See
     // <https://btmc.substack.com/p/how-to-parse-expressions-easy> or <https://tdop.github.io/>
     //
     // If we had any right-associative operators, we would have to use the 'precedence
     // lowering' trick found here: https://btmc.substack.com/p/how-to-parse-expressions-easy
     // Like so: right = parse_expression(tokens, get_precedence(op.tag)-1)
 
-    // TODO I'm going to try and use more combinators/less explicit.
-    // TODO use fn flat_map<G, H, O2>(self, g: G) -> FlatMap<Self, G, O>
-    let (rest, start) = position_ws().parse(input)?;
-    let (rest, op) = parse_binary_operator(rest)?;
-    let (rest, right) = parse_expression(rest, Some(op.thing))?;
-    let (_, end) = nom_locate::position(rest)?;
-    let re = RawExpression::from((left, op, right));
-    Ok((
-        rest,
-        Spanned {
-            span: Span::new(start, end),
-            thing: re,
-        },
-    ))
+    // Note: the spanned combinator can't be used here because `left` was passed to us (thus,
+    // spanned can't capture the start).
+    move |input: ParseInput<'a>| -> ParseResult<Expression> {
```

Comment:
While this is more "elegant", it sure is harder to read for a newbie :\

But this is how we learn; I suspect everything here is really standard Rust. It's just newish to me.

Reply:
But a map over a map is...a brainful.

(BRB gotta look up flat_map vs map)

Reply:
Ok. After reading this, this is a lot less complicated/dramatic a change as I initially feared. LGTM!